File: examples/example_subproblem.jl
using MyOptimizationAlg
using LinearAlgebra
using BenchmarkTools

# Positive definite matrix example
function test_positive_definite()
	println("=== Positive Definite Matrix ===")
	hess = [4.0 1.0; 1.0 3.0]    # Positive definite matrix
	grad = [-1.0, -1.0]          # Gradient vector
	delta = 0.01                  # Trust region radius
	sol = subproblem(hess, grad, delta, verbose = 1)
	println("Solution:", sol)
	println("Norm of solution:", norm(sol))
end

# Non-positive definite matrix example
function test_non_positive_definite()
    println("\n=== Non-Positive Definite Matrix ===")
    hess = [0.0 1.0; 1.0 0.0]    # Not positive definite
    grad = [-1.0, -1.0]          # Gradient vector
    delta = 0.01                  # Trust region radius
    @time sol = subproblem(hess, grad, delta, verbose = 2)
    println("Solution:", sol)
    println("Norm of solution:", norm(sol))
end

# Hard case example
function test_hard_case()
    println("\n=== Hard Case Example ===")
    hess = [2.0 0.0; 0.0 -1.0]   # Indefinite matrix (one negative eigenvalue)
    grad = [-2.0, 1.0]           # Gradient vector
    delta = 1.5                  # Trust region radius (solution must be on the boundary)
    @time sol = subproblem(hess, grad, delta, verbose = 2)
    println("Solution:", sol)
    println("Norm of solution:", norm(sol))
end

# Run the tests
function run_tests()
    test_positive_definite()
    test_non_positive_definite()
    test_hard_case()
end
run_tests()

File: examples/example_subproblem.c
// gcc -Wall -Wextra -Werror -I../src/algorithms example_subproblem.c ../src/algorithms/utilities.c ../src/algorithms/subproblem.c -o example_program -lm

#include "utilities.h"

// Example usage of subproblem subroutine
int main() {
	T hess[N][N] = {0};
	T grad[N] = {0};
	T sol[N] = {0};

	// Fill hess and grad with test values
	for (int i = 0; i < N; i++) {
		grad[i] = i + 1;
		for (int j = 0; j < N; j++) {
			hess[i][j] = (i == j) ? 4.0 : 1.0;
		}
	}

	T delta = 10.0;
	T tol = 1e-4;
	int max_iters = 100;
	int verbose = 0;

	for (int j = 0; j < 100000; j++) {
		subproblem(hess, grad, sol, &delta);
	}

	printf("Solution:\n");
	for (int i = 0; i < N; i++) {
		printf("%.6f ", sol[i]);
	}
	printf("\n");

	return 0;
}



